modules/ipp/scanner.go

```diff
@@ -126,86 +172,394 @@ func (scanner *Scanner) GetPort() uint {
+				//Couldn't read length of content for some reason
+				//return nil, err
+				continue
+			}
+			val := make([]byte, length)
+			if err := binary.Read(buf, binary.BigEndian, &val); err != nil {
+				//Content not read without issue.
+				//vals = append(vals, val)
+				//return vals, err
+				continue
+			}
+			//return &val, nil
+			vals = append(vals, val)
+			if err := binary.Read(buf, binary.BigEndian, &tag); err != nil {
+				//Couldn't read next valueTag
+				//break
+				continue
+			}
+			if err := binary.Read(buf, binary.BigEndian, &nameLength); err != nil {
+				//Couldn't read next nameLength
+				//break
+				// TODO: Conclude correct error behavior for all of these blocks
+				continue
+			}
+		}
+		return vals, nil // FIXME: This should be whatever error occurred during a break case
 	}
-	//FIXME: ?Should just use endpoint "/", since we get the same response as "/ipp" on CUPS??
-	uri := getIPPURL(scanner.config.IPPSecure, host, uint16(scanner.config.BaseFlags.Port), "/ipp")
-	b := getPrinterAttributesRequest(uri)
+	//The attribute was not present
+	return nil, errors.New("Attribute \"" + attrString + "\" not present.")
+}
 
-	resp, err := http.Post(uri, ContentType, &b)
+func versionNotSupported(resp *http.Response, scanner *Scanner) bool {
+	if resp != nil && resp.Body != nil {
+		buf := bufferFromBody(resp, scanner)
+		// Ignore first two bytes, read second two for status code
+		var reader struct {
+			_ uint16
+			StatusCode uint16
+		}
+		err := binary.Read(buf, binary.BigEndian, &reader)
+		if err != nil {
+			return false
+		}
+		// 0x0503 in the second two bytes of the body denotes server-error-version-not-supported
+		// RFC 8011 Section 4.1.8 Source: https://tools.ietf.org/html/rfc8011#4.1.8
+		return reader.StatusCode == 0x0503
+	}
+	return false
+}
+
+func (scanner *Scanner) augmentWithCUPSData(scan *scan, target *zgrab2.ScanTarget, version *version) *zgrab2.ScanError {
+	cupsBody := getPrintersRequest(version.Major, version.Minor)
+	cupsReq, err := http.NewRequest("POST", scan.url, cupsBody)
 	if err != nil {
-		return 0, 0, zgrab2.DetectScanError(err)
+		return zgrab2.NewScanError(zgrab2.SCAN_UNKNOWN_ERROR, err)
+	}
+	cupsReq.Header.Set("Accept", "*/*")
+	cupsReq.Header.Set("Content-Type", ContentType)
+	cupsResp, err := scan.client.Do(cupsReq)
+
+	if versionNotSupported(cupsResp, scanner) {
+		// TODO: Make this step down a version number
+		return nil
+	}
+	scan.results.CUPSResponse = cupsResp
+	if cupsResp != nil && cupsResp.Body != nil {
+		defer cupsResp.Body.Close()
+	}
+	// Store data into BodyText and BodySHA256 of cupsResp
+	storeBody(cupsResp, scanner)
+
+	bodyBytes := []byte(cupsResp.BodyText)
+	cupsVersions, _ := readAttributeFromBody(CupsVersion, &bodyBytes)
+	if len(cupsVersions) > 0 {
+		// Report this cupsVersion b/c it's more detailed than the one found in Server header
+		scan.results.CUPSVersion = strings.Split(scan.results.CUPSVersion, "/")[0] + "/" + string(cupsVersions[0])
```

Comment:
Should these just be different response fields? Maybe `CUPSVersion` is the one from the header, the attribute version could be included alongside other attributes?